### DISCUSSION TITLE
maximum number of connections per node

### DISCUSSION THREAD

**Discussion by lexoring** [xyflow/xyflow#2707] [New Features]
Hello. Now each output of a node has an unlimited number of outputs, or if an input, then it has an unlimited number of inputs to this node.
I need, for example, for one node to limit the number of connections to one. Restriction on the total number of connections to one input or output, some kind of general limit. For example, a node could have a maximum of one connection, and for example, one input could only accept 1 input.
I hope I was able to explain.

![56](https://user-images.githubusercontent.com/74623056/209313111-fbc62831-6fe0-47b7-b0ac-e5d87bbe8492.jpg)

**Comment by bcakmakoglu** [comment]
You can use custom nodes and pass either a `isValidConnection` callback to your handle components inside your custom node (similar to the [Validation](https://reactflow.dev/docs/examples/interaction/validation/) example) or use the `isConnectable` prop of handles to disable connections on this prop altogether (i.e. no crosshair and interactions until the prop is turned to `true` again).

**Reply by lexoring** [reply]
Thanks for the quick response. Honestly, I have been studying the documentation for several days now. It's hard for me to figure it out. I know there is isConnectable which just disables connections, and isValidConnection checks the connection by id.
But I need to have a condition on the number of connections. And I don't even know how to apply these functions to calculate the number of connections. How to find out that the node already has some kind of connection and if there is another user wants to connect, then reject this connection. I don't need to make a condition by id, namely by the number of connections. Thanks for any help.

**Reply by bcakmakoglu** [reply]
Well there is a utility function that is exported by the library called [`getConnectedEdges`](https://reactflow.dev/docs/api/graph-util-functions/#getconnectededges)

You could simply use this, check the returned array for it's length and disable the `isConnectable` prop based on that :)

**Reply by lexoring** [reply]
Thanks for your help!
To be honest, I'm not good at it.
I'm trying to listen for an onConnect event in a node. But such a nuance, when I try to connect to another node, the onconnect event fires, and when from another handle to this one, the event does not work, and thus I cannot check for the maximum number of connections with this handle.

It would be great if there was a handle "maxcount" parameter to simply set how many connections this handle can have. Maybe you can do this. I've been trying to do this for so long and I can't, my experience is not enough. Help me please.

**Reply by lexoring** [reply]
I liked how you did "connectable="single"" for the "handle" in the "view-flow" project
or would it be the same solution as there

`<Handle id="handle-right-a" connectable="single" />`
https://github.com/bcakmakoglu/vue-flow/discussions/370#discussioncomment-3938679

**Reply by bcakmakoglu** [reply]
Alas, React Flow does not offer this option 😅
I implemented it in vue-flow for exactly this case.

[Here's an example of how to do it with `isValidConnection`](https://codesandbox.io/s/rough-mountain-t1qb0r?file=/App.js)
Maybe this helps you :)

**Reply by lexoring** [reply]
Thank you for your help!!! Tell me how to tell the developers to do this thing `<Handle connectable="single" />` how are you or attribute "maxConnections"?
In your example, I can still connect several connections from the single, there seems to be some kind of error, but I can't find it. Could you see this example please.

**Reply by bcakmakoglu** [reply]
I guess the label is misleading, the nodes on the right (A, B, C)  are only connectable with a single edge, not the ones on the left.
If you want them to be single connectable as well, you could update the validator like this

```ts
const useValidatorFn = () => {
  const { getNode, getEdges } = useReactFlow();

  return useCallback(
    (connection) => {
      const edges = getConnectedEdges(
        [getNode(connection.target), getNode(connection.source)],
        getEdges()
      );

      return !edges.length;
    },
    [getNode, getEdges]
  );
};
```

But you could also use a hook to check if the handles are connected and toggle `isConnectable` on the handle.

[Here's another example for that.](https://codesandbox.io/s/confident-tharp-4n3zgy?file=/App.js)

**Reply by lexoring** [reply]
Thank you very much! I will try all of this. It's just that I will have dynamic blocks and it will be difficult for me to track it. There will be different blocks with dynamic IDs. Until then, I'll experiment. 

Merry christmas to you, your help really helped me a lot with react and react-flow. I will hope that they will introduce such a parameter as max connection or single. Thanks

**Reply by khardiman** [reply]
This works well - wrote the following as a common class to be used by all handles in all nodes to ensure that any handle can only be used once:

```
import {useCallback } from 'react';
import { useReactFlow, getConnectedEdges } from 'reactflow';

/**
 * Common usage functions for react-flow
 */
export default class ReactFlowCommon {
    useValidatorFn = () => {
        const { getNode, getEdges } = useReactFlow();
      
        return useCallback(
          (connection) => {
            const edges = getConnectedEdges([getNode(connection.target)], getEdges());
    
            for (let i = 0; i < edges.length; i++) {
              if (edges[i].sourceHandle === connection.sourceHandle || edges[i].targetHandle === connection.targetHandle) {
                console.log('handle already used')
                return false;
              }
            }
            return true;
          },
          [getNode, getEdges]
        );
      };
}
```

referenced with:

`isValidConnection={common.useValidatorFn()}`

after importing.
